## Abstract

We introduce `kmermaid`, a computational method for quantifying similarity RNA-seq transcriptomes across species.
As `kmermaid` skips both alignment and gene orthology assignment, it can be applied to transcriptomes from organisms with no or poorly annotated genomes, and can perform this much faster than current methods without loss of sensitivity.
By enabling analyses across divergent species' transcriptomes in an orthology-, genome- and gene annotation-agnostic manner, `kmermaid` illustrates the potential of non-model organisms in building the cell type evolution tree of life.

<!-- Single-cell RNA-sequencing is a powerful technology for identifying novel and known cell types.
However, the task of identifying new and novel cell types across species is nontrivial, especially when one or more of the species have poorly annotated genomes.
Thus, there is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for a reference genome.
To this end, we have developed a genome-agnostic method to compare molecular profiles using a lossy encoding on $k$-mers from putative protein-coding RNA-seq reads.
We benchmark the annotation-agnostic methods on a bulk comparative transcriptomic dataset consisting of nine species and six tissues, showing that we can recapitulate the results as using only reads mapping to 1:1 orthologs of protein-coding genes, and we are able to resolve ... which can only be seen by using the reference-free $k$-mer method.
We then show that $k$-mers can also be used for comparing transcriptomes built from long read sequencing, by comparing the cell-cell similarity nearest neighbor graphs built on $k$-mers from short reads and long reads from the same cells in a primate brain organoid system.
We find that using $k$-mers on short reads has the advantage of resolving ... in comparison to using read counts from 1:1 gene orthologs, while long reads provide additional information in the form of ...
Using $k$-mers, we were able to resolve cell types X in the primate brain organoid dataset, which was hidden using read counts alone.
Thus, we have show the reference-free methods using the $k$-mers from single cells is a novel, annotation-agnostic method for comparing cells across species that is capable of identifying cell states unique to a particular organism, helping to build the cell type evolution tree of life. -->
