## Abstract

We introduce `kmermaid`, a lightweight computational method for quantifying similarity RNA-seq transcriptomes across species.
As `kmermaid` skips both alignment and gene orthology assignment, it can be applied to transcriptomes from organisms with no or poorly annotated genomes, and can perform this much faster than current methods without loss of sensitivity.
By enabling analyses across divergent species' transcriptomes in an orthology-, genome- and gene annotation-agnostic manner, `kmermaid` illustrates the potential of non-model organisms in building the cell type evolution tree of life.
