## Abstract

We introduce `kmermaid`, a novel computational method for identifying orthologous cell types discovering *de novo* orthologous genes across species.
As `kmermaid` skips both traditional alignment and gene orthology assignment, it can be applied to transcriptomes from organisms with no or poorly annotated genomes, predicts protein-coding sequences from raw RNA-seq reads, and finds commonly expressed predicted protein sequences contributing to shared cell types.
By enabling analyses across divergent species' transcriptomes in an orthology-, genome- and gene annotation-agnostic manner, `kmermaid` illustrates the potential of non-model organisms in building the cell type evolutionary tree of life.
