## Abstract


Single-cell RNA-sequencing is a powerful technology for identifying novel and known cell types.
However, the task of identifying new and novel cell types across species is nontrivial, especially when one or more of the species have poorly annotated genomes.
Thus, there is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for a reference genome.
To this end, we have developed a genome-agnostic method to compare molecular profiles using a lossy encoding on $k$-mers from putative protein-coding RNA-seq reads.
We benchmark the annotation-agnostic methods on a bulk comparative transcriptomic dataset consisting of nine species and six tissues, showing that we can recapitulate the results as using only reads mapping to 1:1 orthologs of protein-coding genes, and we are able to resolve ... which can only be seen by using the reference-free $k$-mer method.
We then show that $k$-mers can also be used for comparing transcriptomes built from long read sequencing, by comparing the cell-cell similarity nearest neighbor graphs built on $k$-mers from short reads and long reads from the same cells.
