## Introduction  {.page_break_before}

There are a predicted 8.7 million Eukaryotic species on earth [@doi:10.1371/journal.pbio.1001127], yet only 14% (1,233,500) have been catalogued and 0.000002% (200/8,700,000 = 2.3e-08) have genomes present in ENSEMBL Assemblies (as of ENSEMBL 98 -- September 2019 release) [@url:https://uswest.ensembl.org/info/website/archives/assembly.html].
And yet, the genome sequence is not enough.
To truly understand the diversity of life on this planet, we need to determine not just the DNA blueprints of life, but understand the instantiation of the DNA, the cell types of the species.
While sequencing DNA gives a quantitative measure of the nucleotide differences, it does not inform the functional strategies that change from DNA modifications due to speciation events .
As new species can be defined by a new cell type.
For example, the existence of a single cell type, the stinging cell called a "Cnidocyte" [@doi:10.1146/annurev-ecolsys-120213-091627], a single-celled biological weapon, defines the phylum Cnidaria.
Thus, entire clades, not only species, can be defined by the introduction of an additional cell type or state.
However, it is unclear how many examples there are of this, and how it is possible to find cell types that are unique to one species.
Thus, we aim to develop a computational method to compare cell types across species that is reference-agnostic and can find cell types that are novel and present in only one organism.

Organizations of existing cell states can also define novel organismal structures.
For example, different physical organizations of similar cell types generate different genitalia in amniotes when comparing mammals to reptiles [@doi:10.1038/nature13819]

Determining common gene ancestry ("orthology") is a difficult problem.

Many approaches exist, reviewed by [@doi:10.1038/nmeth.3830; @doi:10.1093/bioinformatics/btx542; @doi:10.1093/molbev/msz150].
Generally, the approaches are structured in this way: (1) find orthologous groups of genes, (2) build gene trees, (3) build species trees, and (4) assign orthologs, as described in a recent approach (Orthofinder) [@url:https://doi.org/10.1101/466201].
In this approach, we are not interested in exactly reconstructing the species or gene trees, but rather inferring function based on cell type transcriptomes.
Instead of exactly building the gene trees, we subset the protein-coding sequences into peptide words, and re-encode to lossy peptide encodings.


Determining common ancestry of cell types ("orthologous cell types") [@doi:10.1038/nrg.2016.127; @doi:10.1146/annurev-cellbio-100616-060818] is an additional difficult problem.
Comparative transcriptomics begins with finding a common feature set for embedding molecular profiles across divergent species into a common space.
Many researchers take the approach of using one-to-one orthologous genes [@doi:10.1038/nature10532; @doi:10.1038/nbt.4096; @doi:10.1016/j.cell.2019.05.006, @doi:10.1101/459891; @doi:10.1101/397042; @doi:10.1101/371179; @doi:10.1038/s41592-019-0619-0, @url:https://doi.org/10.1101/460246], others use clusters of orthologous groups [@doi:10.1093/nar/28.1.33], others map reads onto a common genome derived from whole-genome alignment [@url:https://doi.org/10.1101/685057; @doi:10.1038/s41586-019-1654-9], or map onto native genomes [@doi:10.1016/j.cell.2019.01.017] and re-annotate using a tool such as Comparative Annotation Toolkit [@doi:10.1101/gr.233460.117].

Annotating one dataset's cell types from another can be performed using random forest models trained on the original dataset [@url:https://doi.org/10.1101/661728], using correlation between cell gene expression profiles as in Cell BLAST [@url:https://doi.org/10.1101/587360], locality-sensitive hashing of bit vectors of gene expression as in CellFishing.jl [@doi:10.1186/s13059-019-1639-x], or using a cell type hierarchy as in Garnett [@doi:10.1038/s41592-019-0535-3] and OnClass [@url:https://doi.org/10.1101/810234].

$k$-mers have been proposed for comparing single cells [@url:https://doi.org/10.1101/723833] as they are a fast, simple way to create cell-cell similarities.
$k$-mers have also been used for orthologous gene detection [@doi:10.1093/gigascience/giz118]
However, the work so far has focused on using annotated organisms and not cross-species analyses.

Reduced amino acid alphabets have been previously used to speed up database searches [@doi:10.1093/bioinformatics/10.4.453; @doi:10.1186/1471-2105-12-159], protein fold prediction [@doi:10.1093/protein/13.3.149; @doi:10.1093/bioinformatics/btp164], and homology recognition [@doi:10.1093/nar/gkh180].
We aim to find "orthologous reads" across species' transcriptomes. 
By representing each species' transcriptome as the set of $k$-mers, we can unbiasedly compare transcriptomes in an orthologous space without the need for knowing the orthologous genes ahead of time, or even the need for a reference genome. 
Additionally, we do not need to reduce the signal to only the genes with a 1:1 orthologous match.
