## Introduction  {.page_break_before}

There are a predicted 8.7 million Eukaryotic species on earth [@doi:10.1371/journal.pbio.1001127], yet only 14% (1,233,500) have been catalogued and 0.000002% (200/8,700,000 = 2.3e-08) have genomes present in ENSEMBL Assemblies (as of ENSEMBL 98 -- September 2019 release) [@url:https://uswest.ensembl.org/info/website/archives/assembly.html].
And yet, the genome sequence is not enough.
To truly understand the diversity of life on this planet, we need to determine not just the DNA blueprints of life, but understand the instantiation of the DNA, the cell types of the species.
While sequencing DNA gives a quantitative measure of the nucleotide differences, it does not inform the functional strategies that change from DNA modifications due to speciation events .
As new species can be defined by a new cell type.
For example, the existence of a single cell type, the stinging cell called a "Cnidocyte" [@doi:10.1146/annurev-ecolsys-120213-091627], a single-celled biological weapon, defines the phylum Cnidaria.
Thus, entire clades, not only species, can be defined by the introduction of an additional cell type or state.

Novel organizations of existing cell states can also define cell types.
For example, the development of genitalia in amniotes, while using similar cell types, ultimately uses a different physical organization of cell types to generate genitalia in mammals compared to reptiles [@doi:10.1038/nature13819]

Determining common gene ancestry ("orthology") is a difficult problem.
Many approaches exist, reviewed by [@doi:10.1038/nmeth.3830; @doi:10.1093/bioinformatics/btx542; @doi:10.1093/molbev/msz150].
Generally, the approaches are structured in this way: (1) find orthologous groups of genes, (2) build gene trees, (3) build species trees, and (4) assign orthologs, as described in a recent approach (Orthofinder) [@url:https://doi.org/10.1101/466201].
In this approach, we are not interested in exactly reconstructing the species or gene trees, but rather inferring function based on cell type transcriptomes.
Instead of exactly building the gene trees, we subset the protein-coding sequences into peptide words, and re-encode to lossy peptide encodings.


Determining common ancestry of cell types ("orthologous cell types") [@doi:10.1038/nrg.2016.127; @doi:10.1146/annurev-cellbio-100616-060818] is an additional difficult problem.
Comparative transcriptomics begins with finding a common feature set for embedding molecular profiles across divergent species into a common space.
Many researchers take the approach of using one-to-one orthologous genes [Cite: brawand2011, CCA, LIGER, Scanorama, basically all the single cell "alignment" packages], others use clusters of orthologous groups [@doi:10.1093/nar/28.1.33], others map reads onto a common genome derived from whole-genome alignment [cite: recent primate brain paper from Barbara Treutlein], or map onto native genomes [@doi:10.1016/j.cell.2019.01.017] and re-annotate using a tool such as Comparative Annotation Toolkit [@doi:10.1101/gr.233460.117].

$k$-mers have been proposed for comparing single cells [@url:https://doi.org/10.1101/723833] as they are a fast, simple way to create cell-cell similarities.
However, the work so far has focused on using annotated organisms and not cross-species analyses.

We aim to find "orthologous reads" across species' transcriptomes. By representing each species' transcriptome as the set of $k$-mers, we can unbiasedly compare transcriptomes in an orthologous space without the need for knowing the orthologous genes ahead of time, or even the need for a reference genome. Additionally, we do not need to reduce the signal to only the genes with a 1:1 orthologous match.
