## Results

### Figure 1 -- $k$-mers are sufficient to detect orthologous genes

To determine whether short segments of sequences could detect gene orthologues, we $k$-merized orthologous genes derived from the ENSEMBL version 97 [@doi:10.1093/nar/gkx1098] COMPARA database [@doi:10.1093/database/bav096] (Figure [@fig:fig1]).
We compared human protein sequences to orthologous chimpanzee, mouse, (orangutan, bonobo, gorilla, macaque, opossum, platypus, chicken) protein sequences, as these are species used in [@doi:10.1038/nature10532].
As a background, we randomly chose 10 non-orthologous genes relative to the human gene.
In addition to $k$-merizing the protein-coding sequence, we also re-encoded the protein-coding sequence into Dayhoff [@raw:dayhoff1969atlas] and hydrophobic-polar encodings [@raw:phillips2012physical], show in Table [@tbl:sequence-encodings].

![**A.** Overview of $k$-mer comparison of orthologous genes. The protein-coding sequence of each pair of known orthologs is $k$-merized, potentially encoded as Dayhoff or Hydrophobic polar, and then the Jaccard index (the intersection divided by the union) is computed on the $k$-mers. **B.** Jaccard similarity of orthologous genes in Dayhoff-encoded $k$-mer space relative to humans in eight species. $x$-axis, $k$-mer size; $y$-axis, Jaccard index. **C.** ](images/figure1.svg){#fig:fig1 width="100%"}

We found that, consistent with previous knowledge, that 1:1 orthologues had higher $k$-mer similarities as determined by the Jaccard Index.

Additionally, more recently diverged genes had higher $k$-mer similarity as well.

#### Outline

- Kmers can approximate orthologies
  - Jaccard similarity of orthologues is higher than non-orthologues
  - Benchmarking using https://orthology.benchmarkservice.org/cgi-bin/gateway.pl
  - Finding orthologues
    - Gold standard
      - ENSEMBL COMPARA
      - Quest for Orthologs consortium, Altenhoff, A. M., Boeckmann, B., Capella-Gutierrez, S., Dalquen, D. A., DeLuca, T., et al. (2016). Standardized benchmarking in the quest for orthologs. Nature Methods, 13(5), 425–430. http://doi.org/10.1038/nmeth.3830 [@doi:10.1038/nmeth.3830]
    - Orthologous groups/Conserved Domain Database [@url:https://www.ebi.ac.uk/miriam/main/collections/MIR:00000119]


### Figure 2 -- $k$-mers from lossily-encoded putative protein-coding reads faithfully pull out reads from protein-coding genes within amniotes

![Figure 2.](images/figure2.svg){#fig:fig2 width="100%"}

- Overview of kmermaid pipeline
  - Comparison of tissue across species
    - Partition reads to coding/noncoding bins
    - MinHash the Dayhoff-encoded coding sequences
    - Jaccard similarity on the MinHashes
- Which reads are found to have coding features but didn’t map to the genome?
- Do these features map to novel genes or gene fusions?
- Kmers can find correct reading from of RNA-seq reads
  - Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data


### Figure 3 -- $k$-mers can pull out only reads from transcription factors and Amniotes can be compared on the MinHashes of their protein-coding sequences

![Figure 3.](images/figure3.svg){#fig:fig3 width="100%"}

- Kmers can find only transcription factor reads of TFs from RNA-seq reads
  - Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data


### Figure 4 -- $k$-mers can compare short and long read datasets in primate brain organoids

![**A.** Overview of experimental system. Poly-A RNA molecules from single cells from primate brain organoids were captured using the Dolomite system, where molecule received a cell barcode and molecular barcode, was reverse transcribed and primed for full-length cDNA. Then the library was split for sequencing on the Illumina or PacBio platforms. For the Illumina platform, the library was first sheared to be compatible with the platform. **B**. UMAP of short read gene counts from Human (left), Chimp (middle), and Orangutan (right) organoids. **C.** UMAP of short read MinHashes with ksize=33, dayhoff encoding and a log2sketchsize=14. **D**. UMAP of long read MinHashes with ksize=33, dayhoff encoding, andl log2sketchsize=14. **E.** Examples of short reads with ambiguous coding sequence resolved by long reads. **F.** UMAP on short-read gene counts of 1:1 orthologous transcription factor genes across all species. **G.** UMAP on short-read MinHashes of all k-mers that match human transcription factor protein-coding peptide sequences. **H.** UMAP on long-read MinHashes of all k-mers that match human transcription factor protein-coding peptide sequences.](images/figure4.svg){#fig:fig4 width="100%"}


References for Primate Brain development

- Florio, M., Heide, M., Pinson, A., Brandl, H., Albert, M., Winkler, S., et al. (2018). Evolution and cell-type specificity of human-specific genes preferentially expressed in progenitors of fetal neocortex. eLife, 7, D635. http://doi.org/10.7554/eLife.32332 [@doi:10.7554/eLife.32332]
- Mazin, P. V., Jiang, X., Fu, N., Han, D., Guo, M., Gelfand, M. S., & Khaitovich, P. (2018). Conservation, evolution, and regulation of splicing during prefrontal cortex development in humans, chimpanzees, and macaques. Rna, 24(4), 585–596. http://doi.org/10.1261/rna.064931.117 [@doi:10.1261/rna.064931.117]
- Xiong, J., Jiang, X., Ditsiou, A., Gao, Y., Sun, J., Lowenstein, E. D., et al. (2018). Predominant patterns of splicing evolution on human, chimpanzee and macaque evolutionary lineages. Human Molecular Genetics, 27(8), 1474–1485. http://doi.org/10.1093/hmg/ddy058 [@doi:10.1093/hmg/ddy058]
